app.py

```diff
@@ -939,38 +982,81 @@ def snap_release(snap_name):
 @app.route('/account/snaps/<snap_name>/market/', methods=['POST'])
 @login_required
 def post_market_snap(snap_name):
-    whitelist = [
-        'title',
-        'summary',
-        'description',
-        'contact_url',
-        'keywords',
-        'license',
-        'price',
-        'blacklist_countries',
-        'whitelist_countries'
-    ]
-
     if 'submit_revert' in flask.request.form:
         flask.flash("All changes reverted.", 'information')
     else:
+        error_list = []
+        info = []
+        files = []
+        images_json = None
+
+        icon = flask.request.files.get('icon')
+        if icon is not None:
+            info.append(build_image_info(icon, 'icon'))
+            files.append(icon)
+
+        screenshots = flask.request.files.getlist('screenshots')
+        for screenshot in screenshots:
+            info.append(build_image_info(screenshot, 'screenshot'))
+            files.append(screenshot)
+
+        if not files:
+            # API requires a multipart request, but we have no files to push
+            # https://github.com/requests/requests/issues/1081
+            files = {'info': ('', dumps(info))}
+        else:
+            images_json = {'info': dumps(info)}
+
+        screenshots_list = snap_screenshots(
+            flask.request.form['snap_id'],
+            images_json,
+            files
+        )
+
+        if 'error_list' in screenshots_list:
+            error_list.append(
+                screenshots_list['error_list']
+            )
+
+        whitelist = [
+            'title',
+            'summary',
+            'description',
+            'contact_url',
+            'keywords',
+            'license',
+            'price',
+            'blacklist_countries',
+            'whitelist_countries'
+        ]
+
         body_json = {
             key: flask.request.form[key]
             for key in whitelist if key in flask.request.form
         }
 
         metadata = snap_metadata(flask.request.form['snap_id'], body_json)
-
         if 'error_list' in metadata:
+            error_list.append(
+                metadata['error_list']
+            )
+
+        if error_list:
+            details = transform_api_data(
+                get_snap_details(snap_name)
+            )
+
             return flask.render_template(
                 'publisher/market.html',
                 snap_id=flask.request.form['snap_id'],
                 snap_name=snap_name,
-                metadata=flask.request.form,
-                error_list=metadata['error_list']
+                **metadata,
```

Comment:
I'm not a fan of implicitly passing attributes through directly from the API. I think the back-end should offer the front-end with certainty about what data it's providing, and raise and handle any errors itself if data is missing. In other words the back-end should be a solid reliable layer in and of itself If you implicitly pass data through then subtle changes in the API data can much more easily cause errors on the frontend because at some point it started relying on data it was never supposed to use.

So I'd rather we grab and explicitly pass through any metadata that the front-end needs. And only what it needs.